standard/ranges.md

```diff
@@ -0,0 +1,309 @@
+> var (ix3, len3) = dropFirst.GetOffsetAndLength(6); // ix3 = 1, len3 = 5
+> var (ix4, len4) = dropLast.GetOffsetAndLength(6);  // ix4 = 0, len4 = 5
+> var (ix5, len5) = maybeLast.GetOffsetAndLength(6); // ix5 = 5, len5 = 1
+> var (ix6, len6) = lastTwo.GetOffsetAndLength(6);   // ix6 = 4, len6 = 2
+> ```
+
+`Range` implements `IEquatable<Range>` and values may be compared for equality based on the abstract value; two `Range` values are equal if and only if the abstract values of the respective `Start` and `End` properties are equal (§24.2). However `Range` values are not ordered and no other comparison operations are provided.
+
+> *Note:* `Range` values are unordered both as they are abstract and there is no unique ordering relation. Once converted to a concrete start and length, e.g. by `GetOffsetAndLength`, an ordering relation could be defined. *end note*
+
+`Range` values can be directly used in the *argument_list* of an *element_access* expression (§12.8.12) which is:
+
+- an array access and the target is a single-dimensional array (§12.8.12.2);
+- a string access (§string-access);
+- an indexer access and the target type has an indexer with corresponding parameters of either `Range` type (§12.8.12.3) or of a type to which `Range` values are implicitly convertible; or
+- an indexer access (§12.8.12.3) and the target type conforms to a sequence pattern for which implicit `Range` support is specified (§24.4.3).
+
+## 24.4 Pattern-based implicit support for Index and Range
+
+### 24.4.1 General
+
+If an *element_access* expression (§12.8.12) of the form `E[A]`; where `E` has type `T` and `A` is a single expression implicitly convertible to `Index` or `Range`; fails to be identified as:
+
+- an array access (§12.8.12.2),
+- a string access (§string-access), or
+- an indexer access (§12.8.12.3) as `T` provides no suitable accessible indexer
+
+then implicit support for the expression is provided if `T` conforms to a particular pattern. If `T` does not conform to this pattern then a compile-time error occurs.
+
+### 24.4.2 Implicit Index support
+
+If in any context an *element_access* expression (§12.8.12) of the form `E[A]`; where `E` has type `T` and `A` is a single expression implicitly convertible to `Index`; is not valid (§24.4.1) then if in the same context:
+
+- `T` provides accessible members qualifying it as a *sequence* (§24.1); and
+- the expression `E[0]` is valid and uses the same indexer that qualifies `T` as a sequence
+
+then the expression `E[A]` shall be implicitly supported.
+
+Without otherwise constraining implementations of this Standard the order of evaluation of the expression shall be equivalent to:
+
+1. `E` is evaluated;
+2. `A` is evaluated;
+3. the countable property of `T` is evaluated, if required by the implementation;
+4. the get or set accessor of the `int` based indexer of `T` that would be used by `E[0]` in the same context is invoked.
```

Comment:
Should this say something about how from-end indexes are evaluated against the countable property to obtain an `int` to pass to the indexer? Same question for slicing below.

Reply:
Details of the calculations are intentionally omitted, note that the countable property need not even be evaluated. The Standard just defines the semantics.